src/core/namespace.c

```diff
@@ -2245,39 +2265,88 @@ int setup_namespace(const NamespaceParameters *p, char **error_path) {
+         * directories, as they are world writable and ephemeral uid/gid will be used. */
+        if (p->private_tmp == PRIVATE_TMP_DISCONNECTED) {
+                _cleanup_free_ char *tmpfs_dir = NULL, *tmp_dir = NULL, *var_tmp_dir = NULL;
+                MountEntry *tmpfs_entry, *tmp_entry, *var_tmp_entry;
 
-                MountEntry *me = mount_list_extend(&ml);
-                if (!me)
+                tmpfs_dir = path_join(p->private_namespace_dir, "unit-private-tmp");
+                tmp_dir = path_join(tmpfs_dir, "tmp");
+                var_tmp_dir = path_join(tmpfs_dir, "var-tmp");
+                if (!tmpfs_dir || !tmp_dir || !var_tmp_dir)
                         return log_oom_debug();
 
-                *me = (MountEntry) {
-                        .path_const = "/tmp",
-                        .mode = ro ? MOUNT_PRIVATE_TMP_READ_ONLY : MOUNT_PRIVATE_TMP,
-                        .source_const = p->tmp_dir,
+                tmpfs_entry = mount_list_extend(&ml);
+                if (!tmpfs_entry)
+                        return log_oom_debug();
+                *tmpfs_entry = (MountEntry) {
+                        .path_malloc = TAKE_PTR(tmpfs_dir),
+                        .mode = MOUNT_PRIVATE_TMPFS,
+                        .options_const = "mode=0700" NESTED_TMPFS_LIMITS,
+                        .flags = MS_NODEV|MS_STRICTATIME,
+                        .has_prefix = true,
                 };
-        }
 
-        if (p->var_tmp_dir) {
-                bool ro = streq(p->var_tmp_dir, RUN_SYSTEMD_EMPTY);
-
-                MountEntry *me = mount_list_extend(&ml);
-                if (!me)
+                tmp_entry = mount_list_extend(&ml);
+                if (!tmp_entry)
                         return log_oom_debug();
+                *tmp_entry = (MountEntry) {
+                        .source_malloc = TAKE_PTR(tmp_dir),
+                        .path_const = "/tmp",
+                        .mode = MOUNT_BIND,
+                        .source_dir_mode = 01777,
+                        .create_source_dir = true,
+                };
 
-                *me = (MountEntry) {
+                var_tmp_entry = mount_list_extend(&ml);
+                if (!var_tmp_entry)
+                        return log_oom_debug();
+                *var_tmp_entry = (MountEntry) {
+                        .source_malloc = TAKE_PTR(var_tmp_dir),
                         .path_const = "/var/tmp",
-                        .mode = ro ? MOUNT_PRIVATE_TMP_READ_ONLY : MOUNT_PRIVATE_TMP,
-                        .source_const = p->var_tmp_dir,
+                        .mode = MOUNT_BIND,
+                        .source_dir_mode = 01777,
+                        .create_source_dir = true,
                 };
+
+                /* Ensure that the tmpfs is mounted first, and bind mounts are added later. */
+                assert_cc(MOUNT_BIND < MOUNT_PRIVATE_TMPFS);
```

Comment:
`mount_path_compare()` first compares entry paths and then mode. So, the check is not sufficient.

Fortunately, the logic works as /run/systemd/unit-private-tmp is earlier than /tmp or /var/tmp.